cluster-autoscaler/core/scale_down.go

```diff
@@ -398,64 +399,79 @@ func (sd *ScaleDown) CleanUpUnneededNodes() {
 // UpdateUnneededNodes calculates which nodes are not needed, i.e. all pods can be scheduled somewhere else,
 // and updates unneededNodes map accordingly. It also computes information where pods can be rescheduled and
 // node utilization level. The computations are made only for the nodes managed by CA.
-// * allNodes are all the nodes processed by CA.
 // * destinationNodes are the nodes that can potentially take in any pods that are evicted because of a scale down.
 // * scaleDownCandidates are the nodes that are being considered for scale down.
-// * pods are the all scheduled pods.
 // * timestamp is the current timestamp.
 // * pdbs is a list of pod disruption budgets.
 func (sd *ScaleDown) UpdateUnneededNodes(
 	destinationNodes []*apiv1.Node,
 	scaleDownCandidates []*apiv1.Node,
 	timestamp time.Time,
-	pdbs []*policyv1.PodDisruptionBudget) errors.AutoscalerError {
+	pdbs []*policyv1.PodDisruptionBudget,
+) errors.AutoscalerError {
 
-	currentlyUnneededNodes := make([]*apiv1.Node, 0)
 	// Only scheduled non expendable pods and pods waiting for lower priority pods preemption can prevent node delete.
-
 	// Extract cluster state from snapshot for initial analysis
-	pods, err := sd.context.ClusterSnapshot.Pods().List(labels.Everything())
-	if err != nil {
-		// This should never happen, List() returns err only because scheduler interface requires it.
-		return errors.ToAutoscalerError(errors.InternalError, err)
-	}
 	allNodeInfos, err := sd.context.ClusterSnapshot.NodeInfos().List()
 	if err != nil {
 		// This should never happen, List() returns err only because scheduler interface requires it.
 		return errors.ToAutoscalerError(errors.InternalError, err)
 	}
 
-	nodeNameToNodeInfo := make(map[string]*schedulernodeinfo.NodeInfo)
+	candidateNames := make(map[string]bool, len(scaleDownCandidates))
+	for _, candidate := range scaleDownCandidates {
+		if candidate == nil {
+			// Do we need to check this?
+			klog.Errorf("Unexpected nil node in node info")
+			continue
+		}
+		candidateNames[candidate.Name] = true
+	}
+	destinationNames := make(map[string]bool, len(destinationNodes))
+	for _, destination := range destinationNodes {
+		if destination == nil {
+			// Do we need to check this?
+			klog.Errorf("Unexpected nil node in node info")
+			continue
+		}
+		destinationNames[destination.Name] = true
+	}
+	candidateNodeInfos := make([]*schedulernodeinfo.NodeInfo, 0, len(candidateNames))
+	destinationNodeInfos := make([]*schedulernodeinfo.NodeInfo, 0, len(destinationNames))
 	for _, nodeInfo := range allNodeInfos {
-		node := nodeInfo.Node()
-		if node != nil {
-			nodeNameToNodeInfo[node.Name] = nodeInfo
+		if nodeInfo.Node() == nil {
+			// Do we need to check this?
+			klog.Errorf("Unexpected nil node in node info")
+			continue
+		}
+		if candidateNames[nodeInfo.Node().Name] {
+			candidateNodeInfos = append(candidateNodeInfos, nodeInfo)
```

Comment:
Probably not a big deal, but previously we had a check for missing NodeInfo. Now if there is a node in candidates list that has no corresponding NI we will just ignore it silently.

Reply:
Good point, will add logging error.